File: frontend/src/lib/api/lidarrRequest.ts
import { api } from '$lib/api/client';

export interface LidarrRequestPayload {
	album_mbid: string;
	track_mbid: string;
	artist_mbid?: string | null;
	track_position?: number | null;
	disc_number?: number | null;
	track_title?: string | null;
}

export interface LidarrRequestAccepted {
	status: string;
	album_id: number;
	album_title: string;
	track_id: number;
	track_title: string;
	other_tracks_unmonitored: number;
	command_id: number | null;
	note: string | null;
}

export interface LidarrRequestTrackStatus {
	recording_mbid: string;
	position: number;
	disc_number: number;
	monitored: boolean;
	has_file: boolean;
}

export interface LidarrRequestStatusResponse {
	in_library: boolean;
	tracks: LidarrRequestTrackStatus[];
}

/** The 3 persistent UI states the LidarrRequestButton can render. */
export type LidarrButtonStatus = 'none' | 'requested' | 'downloaded';

/** Project a single Lidarr per-track entry into a UI-friendly status. */
export function projectButtonStatus(
	t: LidarrRequestTrackStatus | undefined | null
): LidarrButtonStatus {
	if (!t) return 'none';
	if (t.has_file) return 'downloaded';
	if (t.monitored) return 'requested';
	return 'none';
}

/**
 * Build a status lookup over a status response.
 *
 * Indexes by both recording_mbid AND `albumMbid:position:disc` because
 * Lidarr's foreignRecordingId doesn't always equal MusicBrainz's
 * recording_id (Lidarr sometimes maps to a variant). The album_mbid is
 * baked into the position key as a safety belt so a stale lookup from
 * a previous album page can't false-positive-match the visible album's
 * tracks at the same positions (positions 1, 2, 3… collide trivially
 * across every album in existence otherwise).
 */
export function buildStatusLookup(albumMbid: string, res: LidarrRequestStatusResponse) {
	const byMbid = new Map<string, LidarrRequestTrackStatus>();
	const byPositionDisc = new Map<string, LidarrRequestTrackStatus>();
	for (const t of res.tracks) {
		if (t.recording_mbid) byMbid.set(t.recording_mbid, t);
		byPositionDisc.set(`${albumMbid}:${t.position}:${t.disc_number}`, t);
	}
	return {
		albumMbid,
		byMbid,
		byPositionDisc,
		lookup(
			lookupAlbumMbid: string,
			recordingMbid: string | null | undefined,
			position: number,
			discNumber: number
		) {
			if (recordingMbid && byMbid.has(recordingMbid)) return byMbid.get(recordingMbid);
			// Album-scope the position fallback — won't match tracks from a
			// different album even if the caller is somehow using a stale
			// lookup built for that other album.
			if (lookupAlbumMbid !== albumMbid) return undefined;
			return byPositionDisc.get(`${albumMbid}:${position}:${discNumber}`);
		}
	};
}

const ROOT = '/api/v1/lidarr-request';

export async function requestTrackViaLidarr(
	payload: LidarrRequestPayload,
	signal?: AbortSignal
): Promise<LidarrRequestAccepted> {
	return api.global.post<LidarrRequestAccepted>(ROOT, payload, { signal });
}

export async function getLidarrRequestStatus(
	albumMbid: string,
	signal?: AbortSignal
): Promise<LidarrRequestStatusResponse> {
	return api.global.get<LidarrRequestStatusResponse>(
		`${ROOT}/status?album_mbid=${encodeURIComponent(albumMbid)}`,
		{ signal }
	);
}

File: backend/api/v1/schemas/lidarr_request.py
"""Schemas for the inline single-track Lidarr request feature.

Fork-only addition. Distinct from track_download.py (which proxies to
the yt-dlp-worker on gnat for YouTube/Spotify-resolved downloads).

This feature uses Lidarr's native indexer + download-client pipeline
(slskd / qBittorrent / etc.) to grab the requested track, leveraging
the track-monitored fork's PUT /track/monitor endpoint so that only
the requested track ends up on disk — sibling tracks in the same album
release are rejected at import by TrackMonitoredSpecification.

Requires lidarr_url to point at an instance running shaunrd0/Lidarr
(currently only lidarr-shared on gnat:8688).
"""

from __future__ import annotations

from infrastructure.msgspec_fastapi import AppStruct


class LidarrRequestRequest(AppStruct):
    """Inbound request from the frontend.

    album_mbid and track_mbid are required — they drive Lidarr's album
    lookup/add and the per-track identification.

    artist_mbid is optional and only used for diagnostic logging; Lidarr's
    album lookup already returns the artist MBID server-side and that's
    what we actually use for the add. Frontend callers without artist
    context (TopSongsList, etc.) can leave it null.
    """

    album_mbid: str
    track_mbid: str
    artist_mbid: str | None = None
    # Fallback for matching when track_mbid alone doesn't disambiguate.
    # Backend matcher tries them in order: foreignTrackId == track_mbid,
    # foreignRecordingId == track_mbid, position+disc, then track_title.
    # Frontend should send what it has — Popular Songs lists often lack
    # position/disc, album-detail pages have all four.
    track_position: int | None = None
    disc_number: int | None = None
    track_title: str | None = None


class LidarrRequestAccepted(AppStruct):
    """Returned 202 from the request endpoint.

    `command_id` is Lidarr's command queue ID for the AlbumSearch — clients
    can poll `/api/v1/command/{id}` against Lidarr directly if they want
    progress, but for now the UI just shows success/error and lets the
    download appear in the library when Lidarr finishes.
    """

    status: str
    album_id: int
    album_title: str
    track_id: int
    track_title: str
    other_tracks_unmonitored: int
    command_id: int | None = None
    note: str | None = None


class LidarrRequestTrackStatus(AppStruct):
    """Per-track status surfaced by GET /api/v1/lidarr-request/status.

    The frontend uses this to render the LidarrRequestButton in the right
    persistent state — checkmark for downloaded, hourglass for requested-
    but-not-yet-downloaded, idle for not-requested. Without this, the
    button only knows transient session state and forgets after refresh.

    position + disc_number are returned alongside recording_mbid because
    Lidarr's foreignRecordingId doesn't always equal MusicBrainz's
    recording_id (Lidarr sometimes maps to a different variant). The
    frontend prefers recording_mbid match but falls back to position+disc
    when recording_mbid lookup misses.
    """

    recording_mbid: str
    position: int
    disc_number: int
    monitored: bool
    has_file: bool


class LidarrRequestStatusResponse(AppStruct):
    """Returned by GET /api/v1/lidarr-request/status?album_mbid=X.

    in_library = album exists in Lidarr at all. If false, no tracks are
    requested (the button shows idle on every row). If true, walk `tracks`
    to find the per-track status keyed by recording_mbid.
    """

    in_library: bool
    tracks: list[LidarrRequestTrackStatus]
